### PR TITLE
add woo analytics event props for block/shortcode use in cart/checkout

### PR DESCRIPTION
**Pull request by haszari**
This PR adds additional props to `woocommerceanalytics` events so we can filter based on how stores implement their cart & checkout pages.

New [cart](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1289) & [checkout](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1294) blocks are coming soon in the [WooCommerce Blocks plugin](https://github.com/woocommerce/woocommerce-gutenberg-products-block). We're planning to evaluate these blocks by measuring conversion rates from the existing `woocommerceanalytics` events, in particular `add_to_cart`, `product_checkout` and `product_purchase`. To compare traditional and block-based UX, we need to add these new props to the events.

The new props are as follows. Values are `0` or `1`, consistent with other boolean tracks props.

- `cart_page_contains_cart_block`
- `cart_page_contains_cart_shortcode`
- `checkout_page_contains_checkout_block`
- `checkout_page_contains_checkout_shortcode`

Screen Shot 2020-03-26 at 1 02 11 PM

#### Changes proposed in this Pull Request:
* Utility functions for determining if a woo site is using shortcodes or blocks on the cart & checkout pages.
* Include this info as props on all `woocommerceanalytics` events.
* Cache this info in a transient (1-day life, `jetpack-woocommerce-analytics-cart-checkout-info-cache`).

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This adds features to the existing [`woocommerce-analytics`](https://github.com/Automattic/jetpack/tree/master/modules/woocommerce-analytics) module.

#### Testing instructions:
* This requires WooCommerce to be installed and set up on the site, with some products. Recommend setting up `Cash on delivery` payment method for easy checkout.
* The events are disabled for development sites. Use an incognito window, or hack this by adding `return true;` at the start of [`Jetpack_WooCommerce_Analytics::shouldTrackStore()`](https://github.com/Automattic/jetpack/blob/master/modules/woocommerce-analytics/wp-woocommerce-analytics.php#L40).
- Open network tab in browser dev tools, search for `pixel.wp.com`.
- View a product on the front end of your site. You should see a request for `woocommerceanalytics_product_view` event. This should have new properties (parameters) as above. Check that the values are correct. You'll probably have `1` for shortcodes and `0` for blocks.

If you install the current `master` branch of WooCommerce Blocks, you can add the cart & checkout blocks to the cart & checkout pages and retest - the values should show the correct info. If don't have blocks plugin and want to hack some more, you can fake this by adding an [html comment that looks like the start of one of the blocks](https://github.com/Automattic/jetpack/pull/15127/files#diff-0ac5f90177967302e51546ed06d8cca6R471).

You can also test edge cases such as:

- Add the shortcodes and blocks to the pages, i.e. use both at the same time.
- Delete all content from the pages, or switch in some alternative cart or checkout solution (e.g. [One Page Checkout](https://woocommerce.com/products/woocommerce-one-page-checkout/)).
- Add cart/checkout shortcodes/blocks to other pages (not cart/checkout page).
- Break your woo setup or skip onboarding so there is no designated cart/checkout page.

Note that this is cached in a transient for a day, so if you change your setup, you'll need to wait a day to retest, or delete the transient:

`yarn docker:wp transient delete "jetpack-woocommerce-analytics-cart-checkout-info-cache"`

Note: right now, the events are not all consistently firing for block-based cart & checkout (e.g. https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2031). For example, `product_checkout` only works with shortcode checkout. This is a separate issue and will be fixed in a separate PR. Please be aware of this when testing with blocks :)

#### Proposed changelog entry for your changes:
* Added props to WooCommerce Analytics events indicating if store is using blocks or shortcodes on cart & checkout pages.

### CODE REVIEW

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php [hunk 1]
```diff
@@ -416,4 +418,75 @@ public function get_product_categories_concatenated( $product ) {
...
+	 * https://github.com/woocommerce/woocommerce/pull/25932
+	 *
+	 * @param integer $post_id The id of the post to search.
+	 * @param string  $text    The text to search for.
+	 * @return integer 1 if post contains $text (otherwise 0).
+	 */
+	public static function post_contains_text( $post_id, $text ) {
+		global $wpdb;
+
+			// Search for the text anywhere in the post.
```
Comment (jconroy) [inline modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: tiny - whitespace is off

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php [hunk 2]
```diff
@@ -416,4 +418,75 @@ public function get_product_categories_concatenated( $product ) {
...
+	 * whether the store is using shortcodes or Gutenberg blocks.
+	 * This info is cached in a transient.
+	 *
+	 * Note: similar code is in a WooCommerce core PR:
+	 * https://github.com/woocommerce/woocommerce/pull/25932
+	 *
+	 * @return array
+	 */
+	public static function get_cart_checkout_info() {
+		$info = get_transient( 'jetpack-woocommerce-analytics-cart-checkout-info-cache' );
```
Comment (jeherve) [inline modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: Could you switch to using underscores instead of hyphens? It may make things easier to folks searching for all Jetpack transients with `jetpack_`

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php [hunk 3]
```diff
@@ -416,4 +418,75 @@ public function get_product_categories_concatenated( $product ) {
...
+			$info = array(
+				'cart_page_contains_cart_block'         => self::post_contains_text(
+					$cart_page_id,
+					'<!-- wp:woocommerce/cart'
+				),
+				'cart_page_contains_cart_shortcode'     => self::post_contains_text(
+					$cart_page_id,
+					'[woocommerce_cart]'
+				),
+				'checkout_page_contains_checkout_block' => self::post_contains_text(
```
Comment (jeherve) [inline modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: Nitpicky, but I think indenting is off here compared to `checkout_page_contains_checkout_shortcode`.

**Review by jeherve — CHANGES_REQUESTED** [review]
This tests well for me. I only have minor remarks.

**Review by deltaWhiskey — COMMENTED** [review]
I agree with the other reviewers' comments. Otherwise, looks good to me.

I only reviewed code, and did not test.

I'm concerned how this approach will work a year from now. If many features, such as these blocks, will result in properties being added to every `woocommerceanalytics_` event, then the event data could become quite heavy. Especially with the verbose names of the new properties. I do like the self-explanatory names, though.

This is a non-issue if these are the only such properties that will be added in the foreseeable future. But, if more are coming, perhaps the properties should only be added to relevant events. Or a once-per-visit event could be sent with the details for this `url`.

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php (hunk 1 above)
Comment (haszari) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: How'd that sneak through the linter!!

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php (hunk 2 above)
Comment (haszari) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: I prefer that, makes it double-clickable for easy copy-paste too.

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php (hunk 3 above)
Comment (haszari) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: The linter made me do this, I believe it's an esoteric battle between aligning `=>` and keeping lines under 80 chars. Aligning with `checkout_page_contains_checkout_shortcode` would push the other lines out past 80, at least that's what I guessed the linter was thinking!

Happy to change this if needed - let me know specifics.

**Comment by haszari** [comment]
> I'm concerned how this approach will work a year from now. If many features, such as these blocks, will result in properties being added to every woocommerceanalytics_ event, then the event data could become quite heavy. Especially with the verbose names of the new properties. I do like the self-explanatory names, though.
> This is a non-issue if these are the only such properties that will be added in the foreseeable future. But, if more are coming, perhaps the properties should only be added to relevant events. Or a once-per-visit event could be sent with the details for this url.

Good point @deltaWhiskey  - I would definitely prefer to log these more "static" props using such an event. Let's keep this in mind as we iterate - I believe there's no standard mechanism for sending a once-per-visit event right now, so we'd need to develop that.

**Comment by haszari** [comment]
Feedback addressed & rebased the branch - please re-review when ready :)

@jeherve Should I address `codeclimate` issue (below) before merge? I could move `post_contains_text()` and `get_cart_checkout_info()` into a utility class perhaps.

> File wp-woocommerce-analytics-universal.php has 292 lines of code (exceeds 250 allowed). Consider refactoring.

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php (hunk 3 above)
Comment (jeherve) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: Oh, that's interesting. No need to change anything then!

**Comment by jeherve** [comment]
> Should I address codeclimate issue (below) before merge?

No need, I don't think it's really worth it. 

Our CodeClimate configuration is still a work in progress; we need to update its configuration a bit more so it can provide more useful suggestions. Right now it can sometimes be helpful or provide an interesting point of view on how some things could be refactored, but really most of the time it's not super helpful imo. :)

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php [hunk 4]
```diff
@@ -416,4 +418,75 @@ public function get_product_categories_concatenated( $product ) {
...
+					$checkout_page_id,
+					'<!-- wp:woocommerce/checkout'
+				),
+				'checkout_page_contains_checkout_shortcode' => self::post_contains_text(
+					$checkout_page_id,
+					'[woocommerce_checkout]'
+				),
+			);
+
+			set_transient( 'jetpack-woocommerce-analytics-cart-checkout-info-cache', $info, DAY_IN_SECONDS );
```
Comment (jeherve) [inline modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: You'll want to update the transient name here as well, or make it a variable maybe to avoid further issues.

**Review by jeherve — DISMISSED** [review]

File: modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php (hunk 4 above)
Comment (haszari) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: D'oh! Fixed here: 6bfc81731d879bf60f94b4b90dfb931844a6fd9c

Comment (haszari) [reply modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php]: Ready for review again @jeherve :)

**Review by jeherve — APPROVED** [review]
This looks good in my tests. 🚢
